File: log-view-machine/java/src/main/java/com/logview/causality/MachineConfig.java
package com.logview.causality;

import java.util.Collections;
import java.util.Map;
import java.util.Set;

/**
 * Reserved event key for manual/override transitions (e.g. setState without sending an event).
 * Used so every transition has a "cause" (normal event or override); undo-to-root includes overrides.
 */
public final class MachineConfig<T> {

    public static final String OVERRIDE_EVENT = "__override";

    private final String id;
    private final String initial;
    private final T context;
    private final Map<String, StateNodeConfig> states;
    private final Map<String, TransitionTarget> rootOn;

    public MachineConfig(String id, String initial, T context,
                         Map<String, StateNodeConfig> states,
                         Map<String, TransitionTarget> rootOn) {
        this.id = id;
        this.initial = initial;
        this.context = context;
        this.states = states != null ? Map.copyOf(states) : Collections.emptyMap();
        this.rootOn = rootOn != null ? Map.copyOf(rootOn) : null;
    }

    public MachineConfig(String initial, T context, Map<String, StateNodeConfig> states) {
        this(null, initial, context, states, null);
    }

    public String getId() { return id; }
    public String getInitial() { return initial; }
    public T getInitialContext() { return context; }
    public Map<String, StateNodeConfig> getStates() { return states; }
    public Map<String, TransitionTarget> getOn() { return rootOn; }

    /**
     * Resolve transition for event from a state (state's on or root on).
     */
    public String resolveTransition(String currentState, Object event) {
        String eventKey = eventToString(event);
        StateNodeConfig stateConfig = states.get(currentState);
        TransitionTarget transition = null;
        if (stateConfig != null && stateConfig.getOn().containsKey(eventKey)) {
            transition = stateConfig.getOn().get(eventKey);
        }
        if (transition == null && rootOn != null) {
            transition = rootOn.get(eventKey);
        }
        return transition != null ? transition.getTarget() : null;
    }

    String eventToString(Object event) {
        if (event == null) return "null";
        if (event instanceof String) return (String) event;
        if (event instanceof Map<?, ?> m) {
            Object type = m.get("type");
            return type != null ? type.toString() : event.toString();
        }
        return event.toString();
    }

    /**
     * Returns the set of event keys enabled from the given state (state-specific on merged with root on).
     */
    public Set<String> getEnabledEvents(String currentState) {
        StateNodeConfig stateConfig = states.get(currentState);
        Set<String> stateEvents = stateConfig != null ? stateConfig.getOn().keySet() : Set.of();
        Set<String> rootEvents = rootOn != null ? rootOn.keySet() : Set.of();
        return java.util.stream.Stream.concat(stateEvents.stream(), rootEvents.stream())
            .collect(java.util.stream.Collectors.toSet());
    }
}

File: log-view-machine/core/src/main/kotlin/com/logview/causality/MachineConfig.kt
package com.logview.causality

/**
 * Reserved event key for manual/override transitions (e.g. setState without sending an event).
 * Used so every transition has a "cause" (normal event or override); undo-to-root includes overrides.
 */
const val OVERRIDE_EVENT: String = "__override"

/**
 * XState-congruent machine configuration.
 * Matches TS xstateConfig shape: initial, states, context, optional id and global on.
 */
data class TransitionConfig(
    val target: String,
    val actions: List<String>? = null
)

/**
 * Event -> target state (string) or full transition config.
 * In Kotlin we use a sealed representation for on-map values.
 */
sealed class TransitionTarget {
    data class State(val target: String) : TransitionTarget()
    data class WithActions(val target: String, val actions: List<String>) : TransitionTarget()

    companion object {
        fun state(target: String): TransitionTarget = State(target)
    }
}

data class StateNodeConfig(
    val on: Map<String, TransitionTarget> = emptyMap()
)

data class MachineConfig<T>(
    val id: String? = null,
    val initial: String,
    val context: T,
    val states: Map<String, StateNodeConfig> = emptyMap(),
    val on: Map<String, TransitionTarget>? = null
) {
    /** Resolve transition for event from a state (state's on or root on). */
    fun resolveTransition(currentState: String, event: Any): String? {
        val eventKey = eventToString(event)
        val stateConfig = states[currentState]
        val transition = stateConfig?.on?.get(eventKey) ?: on?.get(eventKey) ?: return null
        return when (transition) {
            is TransitionTarget.State -> transition.target
            is TransitionTarget.WithActions -> transition.target
        }
    }

    internal fun eventToString(event: Any): String {
        return when (event) {
            is String -> event
            is Map<*, *> -> (event["type"] as? String) ?: event.toString()
            else -> event.toString()
        }
    }

    /**
     * Returns the set of event keys enabled from the given state (state-specific `on` merged with root `on`).
     * Used to define the n branches of the tri-leaf tree (one per enabled event, plus override).
     */
    fun getEnabledEvents(currentState: String): Set<String> {
        val stateConfig = states[currentState]
        val stateEvents = stateConfig?.on?.keys ?: emptySet<String>()
        val rootEvents = on?.keys ?: emptySet()
        return stateEvents + rootEvents
    }

    fun getInitialContext(): T = context
}
